flink-libraries/flink-table/src/main/scala/org/apache/flink/table/formats/TableFormatFactoryService.scala

```diff
@@ -0,0 +1,245 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.formats
+
+import java.util.{ServiceConfigurationError, ServiceLoader, Map => JMap}
+
+import org.apache.flink.table.api._
+import org.apache.flink.table.descriptors.FormatDescriptorValidator.FORMAT_PROPERTY_VERSION
+import org.apache.flink.table.descriptors._
+import org.apache.flink.table.util.Logging
+
+import _root_.scala.collection.JavaConverters._
+import _root_.scala.collection.mutable
+
+/**
+  * Service provider interface for finding a suitable [[TableFormatFactory]] for the
+  * given properties.
+  */
+object TableFormatFactoryService extends Logging {
+
+  private lazy val defaultLoader = ServiceLoader.load(classOf[TableFormatFactory[_]])
+
+  /**
+    * Finds a table format factory of the given class and creates configured instances from the
+    * given descriptor.
+    *
+    * @param factoryClass desired format factory
```

Comment:
why do we need this param? Shouldn't this be handled by either to instances of this service for serialisation and deserialisation or by copying the providing two sets of methods like `~findSerialisationFactory` and `~findDeserialisationFactory`?

Reply:
As mentioned in the description of this PR, I wanted to make this whole format discovery independent from particular interfaces. Maybe we will introduce block-based writer/reader interfaces or async IO interfaces. This should not require changes in the factory service anymore.